src/rpc/masternode.cpp

```diff
@@ -189,27 +189,37 @@ UniValue masternode(const UniValue& params, bool fHelp)
 
     if (strCommand == "current" || strCommand == "winner")
     {
-        int nCount;
         int nHeight;
+        CScript payee;
+        CTxDestination txdest;
         masternode_info_t mnInfo;
         CBlockIndex* pindex = NULL;
+
         {
             LOCK(cs_main);
             pindex = chainActive.Tip();
         }
         nHeight = pindex->nHeight + (strCommand == "current" ? 1 : 10);
-        mnodeman.UpdateLastPaid(pindex);
 
-        if(!mnodeman.GetNextMasternodeInQueueForPayment(nHeight, true, nCount, mnInfo))
-            return "unknown";
+        if (strCommand == "winner" || !mnpayments.GetBlockPayee(nHeight, payee)) {
+            int nCount;
+            mnodeman.UpdateLastPaid(pindex);
+            if (!mnodeman.GetNextMasternodeInQueueForPayment(nHeight, true, nCount, mnInfo))
+                return "unknown";
+            txdest = mnInfo.pubKeyCollateralAddress.GetID();
+        } else {
+            if (!mnodeman.GetMasternodeInfo(payee, mnInfo))
```

Comment:
It's not clear here that this depends on the result of GetBlockPayee. Any chance you restructure the if/else somehow that the 3 different cases we may have here are more clear? IMHO, in such cases minor code duplication is actually better then having hard to read if/else branches.